Ensure NetworkStorageManager gets destroyed on the main thread

szewai opened the pull request:
#### 64a739754a58011eb565cd42714159be4f9a48f0
<pre>
Ensure NetworkStorageManager gets destroyed on the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=266603">https://bugs.webkit.org/show_bug.cgi?id=266603</a>
<a href="https://rdar.apple.com/117839300">rdar://117839300</a>

Reviewed by Chris Dumez and Alex Christensen.

Dispatch a task to main thread to release ref of NetworkStorageManager in stopReceivingMessageFromConnection, which may
help fix the crash under IDBStorageRegistry::removeConnectionToClient in <a href="https://rdar.apple.com/117839300">rdar://117839300</a>.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):

Canonical link: <a href="https://commits.webkit.org/272313@main">https://commits.webkit.org/272313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3431a81e2f9da7871dc9eb4e7008a39260458290

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31292 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33792 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/28381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12305 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7214 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31628 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8390 "Found 1 new test failure: webrtc/audio-samplerate-change.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27943 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7207 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/7378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35135 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28456 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28297 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33511 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7439 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/5479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31350 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/9114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7348 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8138 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/7962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
```diff
@@ -46,14 +46,18 @@ WebCore::IDBServer::IDBConnectionToClient& IDBStorageRegistry::ensureConnectionT
-        if (connectionToClient->ipcConnection() != connection) {
-            m_connectionsToClient.add(identifier, WTFMove(connectionToClient));
-            continue;
+    m_connectionsToClient.removeIf([&](auto& entry) {
```
The commit log says that this PR also simplifies code in IDBStorageRegistry::removeConnectionToClient(). However, the new code is less concise somehow. Also, it doesn't seem strictly related to the fix. Do we have to do this in this patch?

szewai replied:
The original code creates a temp map from m_connectionsToClient for iteration and adds entries that are not removed back m_connectionsToClient after iteration; the new code just operates on m_connectionsToClient, avoiding move and assignment of values. It's more lengthy as it adds a null check.
This patch aims to fix the crash in the radar (`IDBStorageRegistry::removeConnectionToClient` rdar://117839300), so this change is related.

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
```diff
@@ -46,14 +46,18 @@ WebCore::IDBServer::IDBConnectionToClient& IDBStorageRegistry::ensureConnectionT
+            return true;
+
+        if (connectionToClient->ipcConnection() == connection) {
+            connectionToClient->connectionToClient().connectionToClientClosed();
```
The new code looks more fragile? What if m_connectionsToClient gets modified by connectionToClientClosed() while we're iterating here?

szewai replied:
Updated to the old code.

cdumez commented on Source/WebKit/NetworkProcess/storage/IDBStorageRegistry.cpp:
```diff
@@ -48,6 +48,8 @@ void IDBStorageRegistry::removeConnectionToClient(IPC::Connection::UniqueID conn
 {
     auto allConnectionsToClient = std::exchange(m_connectionsToClient, { });
     for (auto& [identifier, connectionToClient] : allConnectionsToClient) {
+        if (!connectionToClient)
```
We never add null as value, why are we doing this?

szewai replied:
ah you didn't submit the pending comment until just now, that's why I didn't see this. Fixed in the latest patch.

szewai committed:
Ensure NetworkStorageManager gets destroyed on the main thread
https://bugs.webkit.org/show_bug.cgi?id=266603
rdar://117839300

Reviewed by Chris Dumez and Alex Christensen.

Dispatch a task to main thread to release ref of NetworkStorageManager in stopReceivingMessageFromConnection, which may
help fix the crash under IDBStorageRegistry::removeConnectionToClient in rdar://117839300.

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::stopReceivingMessageFromConnection):

Canonical link: https://commits.webkit.org/272313@main
